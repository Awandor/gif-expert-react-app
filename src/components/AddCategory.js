// ========================================
// Functional Component
// ========================================

import { useState } from "react";

import PropTypes from 'prop-types';

const AddCategory = ( { setCategories } ) => {

  const [ inputValue, setInputValue ] = useState( '' ); // Iniciamos con string vacío de lo contrario es undefined y da error

  const handleInputChange = ( event ) => {

    console.log( event.target.value );

    setInputValue( event.target.value ); // sobreescribimos inputValue

  }

  const handleAdd = () => {

    // Para añadir elementos a array hay dos formas

    // setCategories( [...categories, 'Wolverine'] );

    /* setCategories( estadoAnterior =>{

        return [...estadoAnterior, 'Wolverine'];

    } ); */

  };

  const handleSubmit = ( e ) => {

    // No queremos que haga el submit porque eso refresca la página, lo manejaremos de otra forma

    e.preventDefault();

    if ( inputValue.trim().length > 2 ) {

      // Si no tenemos acceso a categories podemos usar callback function que trae categories como argumento

      setCategories( cat => [ inputValue, ...cat ] );

      setInputValue( '' );

    }


  };

  return ( <
    form onSubmit = { handleSubmit } >
    <
    h1 > { inputValue } < /h1> <
    input type = "text"
    value = { inputValue }
    onChange = { handleInputChange }
    /> <
    br / >
    <
    br / >
    <
    button onClick = { handleAdd } > Agregar < /button> <
    /form>
  )

};

AddCategory.propTypes = {
  setCategories: PropTypes.func.isRequired
}

export default AddCategory;
