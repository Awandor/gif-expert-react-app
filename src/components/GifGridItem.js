import React from 'react'

// export const GifGridItem = ( { img } ) => {

export const GifGridItem = ( { id, title, url } ) => {

    // console.log( 'GifGridItem', img );

    return (
        <div className="card animate__animated animate__fadeIn">
           <img src={url} alt={title} />
           <p>{title}</p>
        </div>
    );

};
