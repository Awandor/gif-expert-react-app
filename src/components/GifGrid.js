// import { useEffect, useState } from "react";
import { GifGridItem } from "./GifGridItem";

// import { getGifs } from "../helpers/getGifs";

import { useFetchGifs } from "../hooks/useFetchGifs"


export const GifGrid = ( { category } ) => {

    const { images, loading } = useFetchGifs( category );

    // console.log( loading );
    // console.log( data );

    /* const [images, setImages] = useState( [] )

    useEffect( () => {

        // getGifs( category ).then( resp=>setImages( resp ) ) // Nuevo estado con las imágenes procesadas como queremos);
        getGifs( category ).then( setImages ) // Simplificación

    }, [category] ); // con el arreglo vacío sólo se ejecuta una vez, pero nos sale un warning sobre category */

    return (
        <>
        <h3 className="animate__animated animate__fadeIn">{category}</h3>
        {/* {loading ? 'Cargando...' : 'Datos cargados'} */}
        {loading && <p className="animate__animated animate__flash">Loading</p> }
        <div className="card-grid">
            {
                images.map( ( elem ) => {

                    // key no debe ser el index, debe de ser único, normalmente de base de datos

                    // return <li key={elem.id}>{elem.title}</li>

                    // return <GifGridItem key={elem.id} img={elem} />
                    
                    // Con desestructuración podemos enviar todas las propiedades de elem como un atributo independiente
                    
                    return <GifGridItem key={elem.id} {...elem} />
                
} )
            }
        </div>
        </>
    )

}

// Estamos exportando al comienzo