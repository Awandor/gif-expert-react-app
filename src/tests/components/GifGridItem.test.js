import { shallow } from 'enzyme';
import { GifGridItem } from '../../components/GifGridItem';

// import { render } from '@testing-library/react';

describe( 'Pruebas en <GifGridItem />', () => {

  const id = "werweer34";

  const title = "Algo";

  const url = 'https://algo.com/algo.jpg';

  const wrapper = shallow( <GifGridItem key={id} title={title} url={url} /> );

  // Este test se hace con Enzyme

  test( 'debe mostrar <GifGridItem /> correctamente', () => {

    expect( wrapper ).toMatchSnapshot();

  } );

  test( 'debe tener un parrafo con title', () => {

    const p = wrapper.find( 'p' );

    expect( p.text().trim() ).toBe( title );

  } );
  
  test( 'debe tener una imagen igual al url', () => {

    const img = wrapper.find( 'img' );

    console.log( img.props() ); // props() nos da los atributos de img en forma de propiedades de objeto
    console.log( img.prop( 'src' ) ); // otra manera de acceder a una propiedad

    // expect( img.props().src ).toBe( url );
    expect( img.prop( 'src' ) ).toBe( url );

    expect( img.prop( 'alt' ) ).toBe( title );

  } );

  test( 'debe tener un div con clase animate__fadeIn', () => {

    const div = wrapper.find( 'div' );

    console.log( div.props() );

    expect( div.prop( 'className' ) ).toContain( 'animate__fadeIn' );
    
    expect( div.prop( 'className' ).includes( 'animate__fadeIn' ) ).toBe( true ); // Otra manera

  } );

} );
