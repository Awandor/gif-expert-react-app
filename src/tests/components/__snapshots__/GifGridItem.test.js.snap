// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <GifGridItem /> debe mostrar <GifGridItem /> correctamente 1`] = `
<div
  className="card animate__animated animate__fadeIn"
>
  <img
    alt="Algo"
    src="https://algo.com/algo.jpg"
  />
  <p>
    Algo
  </p>
</div>
`;
